File: Examples/iOS/AudioKitDemo/AudioKitDemo/Processing/ProcessingViewController.m
//
//  ProcessingViewController.m
//  AudioKitDemo
//
//  Created by Aurelius Prochazka on 2/14/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "ProcessingViewController.h"
#import "AKFoundation.h"
#import "ConvolutionInstrument.h"



@implementation ProcessingViewController
{
    IBOutlet UISegmentedControl *sourceSegmentedControl;
    IBOutlet UISwitch *maintainPitchSwitch;
    IBOutlet AKPropertySlider *speedSlider;
    IBOutlet AKPropertySlider *pitchSlider;
    IBOutlet AKPropertySlider *dishWellSlider;
    IBOutlet AKPropertySlider *dryWetSlider;

    float pitchToMaintain;

    ConvolutionInstrument *convolver;
    AKAudioFilePlayer *audioFilePlayer;

    BOOL isPlaying;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    audioFilePlayer = [[AKAudioFilePlayer alloc] init];
    [AKOrchestra addInstrument:audioFilePlayer];

    convolver = [[ConvolutionInstrument alloc] initWithInput:audioFilePlayer.output];
    [AKOrchestra addInstrument:convolver];
    pitchToMaintain = 1.0;
    isPlaying = NO;

    speedSlider.property    = audioFilePlayer.speed;
    pitchSlider.property    = audioFilePlayer.scaling;
    dishWellSlider.property = convolver.dishWellBalance;
    dryWetSlider.property   = convolver.dryWetBalance;

}

- (void)viewWillDisappear:(BOOL)animated   {
    [super viewWillDisappear:animated];
    [self stop:self];
}


- (IBAction)start:(id)sender {
    if (!isPlaying) {
        [convolver start];
        [audioFilePlayer play];
        isPlaying = YES;
    }
}

- (IBAction)stop:(id)sender {
    if (isPlaying) {
        [audioFilePlayer stop];
        [convolver stop];
        isPlaying = NO;
    }

}


- (IBAction)speedChanged:(UISlider *)sender
{
    if (maintainPitchSwitch.isOn && fabs(audioFilePlayer.speed.value) > 0.1) {
        audioFilePlayer.scaling.value =  pitchToMaintain / fabs(audioFilePlayer.speed.value);
    }
}

- (IBAction)togglePitchMaintenance:(UISwitch *)sender {
    if (sender.isOn) {
        [pitchSlider setEnabled:NO];
        pitchToMaintain = fabs(audioFilePlayer.speed.value) * audioFilePlayer.scaling.value;
    } else {
        [pitchSlider setEnabled:YES];
    }
}
- (IBAction)fileChanged:(UISegmentedControl *)sender {
    audioFilePlayer.sampleMix.value = (float) sender.selectedSegmentIndex;
}

@end

File: Examples/iOS/Swift/AudioKitDemo/AudioKitDemo/Processing/ProcessingViewController.swift
//
//  ProcessingViewController.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//


class ProcessingViewController: UIViewController {

    @IBOutlet var sourceSegmentedControl: UISegmentedControl!
    @IBOutlet var maintainPitchSwitch: UISwitch!
    @IBOutlet var speedSlider: AKPropertySlider!
    @IBOutlet var pitchSlider: AKPropertySlider!
    @IBOutlet var dishWellSlider: AKPropertySlider!
    @IBOutlet var dryWetSlider: AKPropertySlider!

    var isPlaying = false

    var pitchToMaintain: Float = 1.0

    var conv: ConvolutionInstrument!
    let audioFilePlayer = AKAudioFilePlayer()

    override func viewDidLoad() {
        super.viewDidLoad()

        conv = ConvolutionInstrument(input: audioFilePlayer.output)

        AKOrchestra.addInstrument(audioFilePlayer)
        AKOrchestra.addInstrument(conv)

        speedSlider.property = audioFilePlayer.speed
        pitchSlider.property = audioFilePlayer.scaling
        dryWetSlider.property   = conv.dryWetBalance
        dishWellSlider.property = conv.dishWellBalance

    }


    @IBAction func start(sender:UIButton) {
        if (!isPlaying) {
            conv.play()
            audioFilePlayer.play()
            isPlaying = true
        }
    }

    @IBAction func stop(sender:UIButton) {
        if (isPlaying) {
            conv.stop()
            audioFilePlayer.stop()
            isPlaying = false
        }
    }

    @IBAction func speedChanged(sender:UISlider) {
        if (maintainPitchSwitch.on && fabs(audioFilePlayer.speed.value) > 0.1) {
            audioFilePlayer.scaling.value = pitchToMaintain / fabs(audioFilePlayer.speed.value)
        }
    }


    @IBAction func togglePitchMaintenance(sender:UISwitch) {
        if sender.on {
            pitchSlider.enabled = false
            pitchToMaintain = fabs(audioFilePlayer.speed.value) * audioFilePlayer.scaling.value
        } else {
            pitchSlider.enabled = true
        }
    }

    @IBAction func fileChanged(sender:UISegmentedControl) {
        audioFilePlayer.sampleMix.value = Float(sender.selectedSegmentIndex)
    }
}
